File: Leetcode/0239 Sliding Window Maximum.py
class Solution:
    def maxSlidingWindow(self, nums, k):
        sliding_window_max = []
        decresing_queue = deque()
        for i, num in enumerate(nums):
            while decresing_queue and nums[decresing_queue[-1]] < num:
                decresing_queue.pop()

            decresing_queue.append(i)

            if (i + 1) < k:
                continue

            exiting_index = i - k
            if 0 <= exiting_index and decresing_queue[0] == exiting_index:
                decresing_queue.popleft()

            sliding_window_max.append(nums[decresing_queue[0]])

        return sliding_window_max

File: Leetcode/0239 Sliding Window Maximum.java
public class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] slidingWindowMax = new int[nums.length - k];
        int slidingWindowIndex = 0;

        Deque<Integer> decreasingDeque = new LinkedList<>();

        for (int i = 0; i < nums.length; ++i) {
            while (!decreasingDeque.isEmpty() && decreasingDeque.peekLast() < nums[i]) {
                decreasingDeque.pollLast();
            }
            decreasingDeque.addLast(i);

            if (i + 1 < k) {
                continue;
            }

            int exitingIndex = i - k;
            if (0 <= exitingIndex && decreasingDeque.peekFirst() == exitingIndex) {
                decreasingDeque.pollFirst();
            }
            slidingWindowMax[slidingWindowIndex++] = nums[decreasingDeque.peekFirst()];
        }
        return slidingWindowMax;
    }
}